[SPARK-13659] Refactor BlockStore put*() APIs to remove returnValues

JoshRosen opened the pull request:
In preparation for larger refactoring, this patch removes the confusing `returnValues` option from the BlockStore put() APIs: returning the value is only useful in one place (caching) and in other situations, such as block replication, it's simpler to put() and then get().

As part of this change, I needed to refactor `BlockManager.doPut()`'s block replication code. I also changed `doPut()` to access the memory and disk stores directly rather than calling them through the BlockStore interface; this is in anticipation of a followup patch to remove the BlockStore interface so that the disk store can expose a binary-data-oriented API which is not concerned with Java objects or serialization.

These changes should be covered by the existing storage unit tests. The best way to review this patch is probably to look at the individual commits, all of which are small and have useful descriptions to guide the review.

/cc @davies for review.

JoshRosen committed:
Refactor synchronous block replication code to call doGetLocal():

The goal behind this change is to rely on doGetLocal() for returning bytes
rather than having to perform the serialization in the replication code.

JoshRosen committed:
Don't request returnValues when putting values in doPut().

JoshRosen committed:
Remove unused PutResult.droppedBlocks field

JoshRosen committed:
Clarify that putBytes() never has returnValues == true

JoshRosen committed:
Don't call putIterator() with returnValues == true in tests.

JoshRosen committed:
Remove returnValues from the BlockStore interface.

JoshRosen committed:
Change putIterator return type to be an Either.

JoshRosen committed:
Remove last accesses of PutResult.data.

JoshRosen committed:
Remove the PutResult.data field.

JoshRosen committed:
Access DiskStore and MemoryStore directly instead of through BlockStore interface in doPut().

JoshRosen committed:
Remove unused DiskStore.getBytes() overload.

JoshRosen committed:
Change putBytes' return type to Unit, since it's never checked.

Eventually it would make sense to give this a Boolean return type, but
I'd like to do this later since there's actually a tricky semantics issue
when considering what the return type is when saving to the MemoryStore.

JoshRosen committed:
Remove PutResult.

SparkQA commented:
**[Test build #52419 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52419/consoleFull)** for PR 11502 at commit [`6381b00`](https://github.com/apache/spark/commit/6381b00a94c7bf4ea0693fc4ae6868ef0f866dc4).

davies commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -744,7 +750,7 @@ private[spark] class BlockManager(
       tellMaster: Boolean = true,
       effectiveStorageLevel: Option[StorageLevel] = None,
-      keepReadLock: Boolean = false): Option[PutResult] = {
+      keepReadLock: Boolean = false): Option[Option[Iterator[Any]]] = {
```
Do we really need this two level options?

andrewor14 replied:
I checked the callsites. Currently we never actually use the inner level of option. The only place where we look at the `Some` value is in `getOrElseUpdate`. There we just do `x.get.get` anyway, so this can just be an `Option[Iterator[Any]]`.

If there's another boolean that we need to return, then I would just make a private case class even if it's called in only 1 place. Right now it's unclear what `Some(None)` means, for example. Anything we can think of here is better than nested options.

JoshRosen replied:
`doPut()` can be called to write either bytes or an iterator of values (since `doPut` is used in the implementation of both `putIterator` and `putBytes`). 

Imagine that `putBytes()` fails because there was not enough space to store the block in memory and it could not be dropped to disk. In this case, the `put()` call has failed but we can't return an iterator.

Thus, we need to make a case class so that we can convey the three separate outcomes (put of bytes failed, put of iterator failed, and put succeeded or block was already present).

davies commented:
I'm not familiar with BlockManager, the changes looks straightforward to me, only one minor question.

LGTM over all.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -734,9 +740,9 @@ private[spark] class BlockManager(
-   * @return `Some(PutResult)` if the block did not exist and could not be successfully cached,
-   *         or None if the block already existed or was successfully stored (fully consuming
-   *         the input data / input iterator).
+   * @return `Some(Option[Iterator[Any])` if the block did not exist and could not be successfully
```
I haven't done a full review yet but this return type is super confusing...

JoshRosen replied:
Basically, the problem is that we need to distinguish between three outcomes: success, failure while storing bytes, and failure while storing an iterator. This is only called in one place internally, so I got lazy and neglected to make a simpler return type to capture this distinction.

JoshRosen replied:
This will be much easier to clean up if we split the monolithic `doPut()` into a `doPutBytes()` and `doPutIterator()`.

davies replied:
Sounds good to me.

SparkQA commented:
**[Test build #52419 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52419/consoleFull)** for PR 11502 at commit [`6381b00`](https://github.com/apache/spark/commit/6381b00a94c7bf4ea0693fc4ae6868ef0f866dc4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52419/
Test FAILed.

JoshRosen commented:
I think this failed an unrelated flaky test in `CommitFailureTestRelationSuite`.

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #52438 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52438/consoleFull)** for PR 11502 at commit [`6381b00`](https://github.com/apache/spark/commit/6381b00a94c7bf4ea0693fc4ae6868ef0f866dc4).

SparkQA commented:
**[Test build #52438 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52438/consoleFull)** for PR 11502 at commit [`6381b00`](https://github.com/apache/spark/commit/6381b00a94c7bf4ea0693fc4ae6868ef0f866dc4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52438/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #52478 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52478/consoleFull)** for PR 11502 at commit [`6381b00`](https://github.com/apache/spark/commit/6381b00a94c7bf4ea0693fc4ae6868ef0f866dc4).

SparkQA commented:
**[Test build #52478 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52478/consoleFull)** for PR 11502 at commit [`6381b00`](https://github.com/apache/spark/commit/6381b00a94c7bf4ea0693fc4ae6868ef0f866dc4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52478/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #52489 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52489/consoleFull)** for PR 11502 at commit [`6381b00`](https://github.com/apache/spark/commit/6381b00a94c7bf4ea0693fc4ae6868ef0f866dc4).

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -432,98 +432,104 @@ private[spark] class BlockManager(
+    }
+  }
+
+  private def doGetLocal(
```
this assumes you already hold the read lock for the block. We should explicitly say so in a comment

JoshRosen replied:
Yep, can do.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -801,43 +798,45 @@ private[spark] class BlockManager(
+            case IteratorValues(iterator) =>
+              diskStore.putIterator(blockId, iterator(), putLevel) match {
+                case Right(s) =>
+                  size = s
```
can this never be `Left(iter)`? I think we should still match it to throw a better exception than "match error". If the concern is that we will remove `BlockStore` in a future patch then we should add a TODO here instead.

JoshRosen replied:
Nope, it can't; the return type of `putIterator` is actually `Right` (check the signature). That's why this isn't complaining about a partial match.

JoshRosen replied:
I didn't bother to comment this because this code is going away in my followup patch; the weird type will no longer be necessary once MemoryStore and DiskStore no longer share a common superclass.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -868,34 +867,23 @@ private[spark] class BlockManager(
+      val remoteStartTime = System.currentTimeMillis
+      val bytesToReplicate = doGetLocal(blockId, putBlockInfo, asBlockResult = false).getOrElse {
+        throw new SparkException(s"Block $blockId was not found even though it was just stored")
+      }.asInstanceOf[ByteBuffer]
```
style: slightly prefer
```
doGetLocal(...)
  .map(_.asInstanceOf[ByteBuffer])
  .getOrElse { throw new SparkException(...) }
```

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -734,9 +740,9 @@ private[spark] class BlockManager(
-   *         the input data / input iterator).
+   * @return `Some(Option[Iterator[Any])` if the block did not exist and could not be successfully
+   *          cached, or None if the block already existed or was successfully stored (fully
+   *          consuming the input data / input iterator).
```
this paragraph is a little too low level. I would start by explaining what the return value means on a high level, e.g.
```
@return iterator of values if the put failed. This is defined only if ...
```
then go into what `Some` or `None` means.

SparkQA commented:
**[Test build #52489 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52489/consoleFull)** for PR 11502 at commit [`6381b00`](https://github.com/apache/spark/commit/6381b00a94c7bf4ea0693fc4ae6868ef0f866dc4).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52489/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #52499 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52499/consoleFull)** for PR 11502 at commit [`6381b00`](https://github.com/apache/spark/commit/6381b00a94c7bf4ea0693fc4ae6868ef0f866dc4).

SparkQA commented:
**[Test build #52499 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52499/consoleFull)** for PR 11502 at commit [`6381b00`](https://github.com/apache/spark/commit/6381b00a94c7bf4ea0693fc4ae6868ef0f866dc4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52499/
Test PASSed.

JoshRosen committed:
Address review comments.

JoshRosen commented:
Alright, I've updated this to address the review comments. Note, however, that most of the code that I changed in my latest commit is going to be immediately deleted / rewritten in the larger refactoring patch that will follow this one.

SparkQA commented:
**[Test build #52503 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52503/consoleFull)** for PR 11502 at commit [`3d8f479`](https://github.com/apache/spark/commit/3d8f47931915e39feea45f4be30280bb7b472a4b).

JoshRosen committed:
Comment typo.

SparkQA commented:
**[Test build #52504 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52504/consoleFull)** for PR 11502 at commit [`910f997`](https://github.com/apache/spark/commit/910f9977e5537410da994a0e93096dd7e1e8dbfe).

SparkQA commented:
**[Test build #52503 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52503/consoleFull)** for PR 11502 at commit [`3d8f479`](https://github.com/apache/spark/commit/3d8f47931915e39feea45f4be30280bb7b472a4b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52503/
Test PASSed.

SparkQA commented:
**[Test build #52504 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52504/consoleFull)** for PR 11502 at commit [`910f997`](https://github.com/apache/spark/commit/910f9977e5537410da994a0e93096dd7e1e8dbfe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52504/
Test PASSed.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -54,15 +54,12 @@ private[spark] class DiskStore(blockManager: BlockManager, diskManager: DiskBloc
-      level: StorageLevel,
-      returnValues: Boolean): PutResult = {
-
+      level: StorageLevel): Right[Iterator[Any], Long] = {
```
This return type is really wonky. We can fix this later though.

andrewor14 commented:
LGTM I'm merging this into master.

pzz2011 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -432,98 +432,105 @@ private[spark] class BlockManager(
+    // Look for the block in memory
+    if (level.useMemory) {
+      logDebug(s"Getting block $blockId from memory")
+      val result = if (asBlockResult) {
```
Hey? excuse me, @JoshRosen  can I ask you a question(may be a little stupid...)? what's the `asBlockResult` mean? why here should use it?
